orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -133,7 +177,19 @@ Node* AddBackwardRecord(Graph& graph,
                       std::begin(backward_send->MutableOutputDefs()),
                       std::end(backward_send->MutableOutputDefs()));
   }
-  FindLeafNodes(graph, input_args);
+
+  // Find all leaf nodes' frist inputs. They are used togehter as control edges
+  // to determine if backward pass is finished.
+  auto backward_leaf_node_args = FindBackwardLeafNodes(graph);
+
+  // For each leaf tensor in the backward pass, we use "Group" operator to
+  // convert it to a boolean scalar so that the original leaf's memory can be
+  // released earlier.
+
+  // TODO: use full list instead of the first element after changining
+  // topological sort to depth-first from inputs.
+  std::vector<NodeArg*> sub_backward_leaf_node_args{backward_leaf_node_args[0]};
+  ConvertTensorToBoolSignal(graph, backward_leaf_node_args, input_args);
```

Comment:
>backward_leaf_node_args [](start = 35, length = 23)

Should this be sub_backward_leaf_node_args?

Reply:
Ah.....yes.

---
In reply to: [439701645](https://github.com/microsoft/onnxruntime/pull/4204#discussion_r439701645) [](ancestors = 439701645)